content/300-guides/100-performance-and-optimization/090-metrics.mdx

```diff
@@ -0,0 +1,477 @@
+pool_wait_duration_ms_bucket{le="+Inf"} 3
+pool_wait_duration_ms_sum 0.023208
+pool_wait_duration_ms_count 3
+
+# HELP query_operation_total_elapsed_time_ms
+# TYPE query_operation_total_elapsed_time_ms histogram
+query_operation_total_elapsed_time_ms_bucket{le="0"} 0
+query_operation_total_elapsed_time_ms_bucket{le="1"} 1
+query_operation_total_elapsed_time_ms_bucket{le="5"} 2
+query_operation_total_elapsed_time_ms_bucket{le="10"} 2
+query_operation_total_elapsed_time_ms_bucket{le="50"} 2
+query_operation_total_elapsed_time_ms_bucket{le="100"} 2
+query_operation_total_elapsed_time_ms_bucket{le="500"} 2
+query_operation_total_elapsed_time_ms_bucket{le="1000"} 2
+query_operation_total_elapsed_time_ms_bucket{le="5000"} 2
+query_operation_total_elapsed_time_ms_bucket{le="50000"} 2
+query_operation_total_elapsed_time_ms_bucket{le="+Inf"} 2
+query_operation_total_elapsed_time_ms_sum 3.197624
+query_operation_total_elapsed_time_ms_count 2
+
+# HELP query_total_elapsed_time_ms
+# TYPE query_total_elapsed_time_ms histogram
+query_total_elapsed_time_ms_bucket{le="0"} 0
+query_total_elapsed_time_ms_bucket{le="1"} 5
+query_total_elapsed_time_ms_bucket{le="5"} 5
+query_total_elapsed_time_ms_bucket{le="10"} 5
+query_total_elapsed_time_ms_bucket{le="50"} 5
+query_total_elapsed_time_ms_bucket{le="100"} 5
+query_total_elapsed_time_ms_bucket{le="500"} 5
+query_total_elapsed_time_ms_bucket{le="1000"} 5
+query_total_elapsed_time_ms_bucket{le="5000"} 5
+query_total_elapsed_time_ms_bucket{le="50000"} 5
+query_total_elapsed_time_ms_bucket{le="+Inf"} 5
+query_total_elapsed_time_ms_sum 1.8407059999999997
+query_total_elapsed_time_ms_count 5
+```
+
+### Use Prisma metrics with the Prometheus metrics system
+
+In the majority of cases, Prometheus must scrape an endpoint to retrieve metrics. The following example shows how to send data with `Express.js`:
+
+```js
+import { PrismaClient } from "@prisma/client";
+import express, { Request, Response } from "express";
+
+const app = express();
+const port = 4000;
+const prisma = new PrismaClient();
+
+app.get("/metrics", async (\_req, res: Response) => {
+  let metrics = await prisma.$metrics.prometheus();
+  res.end(metrics);
+});
+
+app.listen(port, () => {
+  console.log(`Example app listening on port ${port}`);
+})
+```
+
+The following example shows how to combine Prisma metrics with other Prometheus client libraries, also served via an `Express.js` endpoint:
+
+```js
+import { PrismaClient } from "@prisma/client";
+import express, { Request, Response } from "express";
+import prom from "prom-client";
+
+let registry = new prom.Registry();
+const app = express();
+const port = 4000;
+const prisma = new PrismaClient();
+
+app.get("/metrics", async (\_req, res: Response) => {
+let prismaMetrics = await prisma.$metrics.prometheus();
+  let appMetrics = await registry.metrics();
+  res.end(prismaMetrics + appMetrics);
+});
+
+app.listen(port, () => {
+  console.log(`Example app listening on port ${port}`);
+});
+```
```

Comment:
I tried running this code, but the `appMetrics` from the other registry is empty, which was a bit confusing. 

Suggestion: perhaps it makes sense to show an example that exposes actual metrics besides ones from Prisma? 

The example can be tweaked a bit to use the default metrics exposed by `prom-client`. 

```suggestion
```ts
import { PrismaClient } from "@prisma/client";
import express, { Request, Response } from "express";
import client from "prom-client";

const app = express();
const port = 4000;
const prisma = new PrismaClient();

const register = new client.Registry()
client.collectDefaultMetrics({ register })

app.get("/metrics", async (_req, res: Response) => {
  let prismaMetrics = await prisma.$metrics.prometheus();
  let appMetrics = await register.metrics();

    res.end(prismaMetrics + appMetrics);
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`);
});
```
```